File: solutions/2026-02-20/javascript/2026-02-20-special-binary-string.js
/**
 * @summary LeetCode 761: Special Binary String. This problem involves rearranging a special binary string by swapping adjacent special substrings to achieve the lexicographically largest result. Special binary strings have an equal number of 0s and 1s, and every prefix has at least as many 1s as 0s.
 * @link https://leetcode.com/problems/special-binary-string/
 * @approach The core idea is to recursively break down the special binary string into its smallest special components. A special binary string can be uniquely represented as "1" + S1 + "0" + S2, where S1 and S2 are themselves special binary strings (potentially empty). The recursive structure allows us to sort these inner special strings lexicographically. We can find these components by scanning the string, maintaining a balance counter (increment for '1', decrement for '0'). When the balance returns to zero, we've found a complete special substring. We recursively process the inner parts and then sort them in descending lexicographical order. The final result is constructed by concatenating "1", the sorted inner strings, "0", and then recursively processing the remaining part of the string.
 * @time_complexity The time complexity is roughly O(N * log N) in the best case due to sorting, but can approach O(N^2) in the worst case due to string concatenations and recursive calls on substrings. Given the small constraint (N <= 50), this is acceptable.
 * @space_complexity The space complexity is O(N) due to the recursion depth and the creation of new strings during processing.
 */
var makeLargestSpecial = function(s) {
    // Base case: if the string is empty or invalid, return it.
    if (!s || s.length === 0) {
        return "";
    }

    // We will store the special components found within the current string.
    const components = [];
    let balance = 0; // Tracks the balance of '1's and '0's.
    let start = 0;   // Marks the beginning of a potential special substring.

    // Iterate through the string to identify and extract special components.
    for (let i = 0; i < s.length; i++) {
        if (s[i] === '1') {
            balance++; // Increment balance for '1'
        } else {
            balance--; // Decrement balance for '0'
        }

        // When balance returns to 0, we have found a complete special substring.
        if (balance === 0) {
            // The special substring is of the form "1" + inner_special_string + "0".
            // The inner_special_string is the substring from start + 1 to i - 1.
            // Recursively call makeLargestSpecial on the inner part to ensure it's also lexicographically largest.
            components.push("1" + makeLargestSpecial(s.substring(start + 1, i)) + "0");
            // Update the start for the next component.
            start = i + 1;
        }
    }

    // Sort the extracted special components in descending lexicographical order.
    // This is crucial for achieving the overall lexicographically largest string.
    components.sort((a, b) => b.localeCompare(a));

    // Concatenate the sorted components to form the result.
    return components.join("");
};
```

File: solutions/2026-02-20/python/2026-02-20-special-binary-string.py
```python
# Problem: Special Binary String
# Summary: Rearrange a special binary string by swapping consecutive special substrings to get the lexicographically largest result.
# Link: https://leetcode.com/problems/special-binary-string/
#
# Approach:
# This problem can be solved using a recursive divide and conquer approach.
# A special binary string can be broken down into smaller special binary strings.
# The core idea is that any special binary string can be represented as '1' followed by zero or more special binary strings, followed by '0'.
# For example, "11011000" can be seen as "1" + "101100" + "0". The inner "101100" can be further decomposed.
# We can find these components by maintaining a balance counter (increment for '1', decrement for '0'). When the balance returns to 0, we've found a complete special substring.
#
# The key insight for lexicographical ordering is that within a larger special string formed by "1" + inner_special_strings + "0", we want to arrange the `inner_special_strings` to be as large as possible.
# This means we should recursively sort the `inner_special_strings` themselves and then concatenate them in descending order before appending the final '0'.
#
# Algorithm:
# 1. Define a recursive function `makeLargestSpecial(s)`.
# 2. Initialize an empty list `components` to store the special substrings.
# 3. Initialize `balance = 0` and `start_index = 0`.
# 4. Iterate through the string `s` with index `i`:
#    a. If `s[i] == '1'`, increment `balance`.
#    b. If `s[i] == '0'`, decrement `balance`.
#    c. If `balance == 0`:
#       i. This marks the end of a special substring from `start_index` to `i`.
#       ii. The substring is `s[start_index : i + 1]`.
#       iii. If this substring is "10", add it directly to `components`.
#       iv. Otherwise, it's of the form "1" + inner_special + "0".
#           - Recursively call `makeLargestSpecial` on the inner part: `s[start_index + 1 : i]`.
#           - Prepend '1' and append '0' to the result of the recursive call to form the largest special substring from this segment.
#           - Add this formed largest special substring to `components`.
#       v. Update `start_index = i + 1`.
# 5. After iterating through the string, sort the `components` list in reverse lexicographical order.
# 6. Join the sorted `components` to form the final result string.
#
# Example Walkthrough: s = "11011000"
#
# makeLargestSpecial("11011000")
#   i=0, s[0]='1', balance=1
#   i=1, s[1]='1', balance=2
#   i=2, s[2]='0', balance=1
#   i=3, s[3]='1', balance=2
#   i=4, s[4]='1', balance=3
#   i=5, s[5]='0', balance=2
#   i=6, s[6]='0', balance=1
#   i=7, s[7]='0', balance=0. Found a special substring "11011000" (from index 0 to 7).
#     This is "1" + "101100" + "0".
#     Recursively call makeLargestSpecial("101100").
#
#       makeLargestSpecial("101100")
#         i=0, s[0]='1', balance=1
#         i=1, s[1]='0', balance=0. Found "10". Add "10" to components. start_index=2.
#         i=2, s[2]='1', balance=1
#         i=3, s[3]='1', balance=2
#         i=4, s[4]='0', balance=1
#         i=5, s[5]='0', balance=0. Found "1100". This is "1" + "10" + "0".
#           Recursively call makeLargestSpecial("10").
#
#             makeLargestSpecial("10")
#               i=0, s[0]='1', balance=1
#               i=1, s[1]='0', balance=0. Found "10". Add "10" to components. start_index=2.
#             Return "10".
#
#           Append "1", result_of_rec("10"), "0" -> "1" + "10" + "0" = "1100". Add "1100" to components.
#         Components for "101100" are ["10", "1100"].
#         Sort reverse: ["1100", "10"].
#         Join: "110010".
#       Return "110010".
#
#     Append "1", result_of_rec("101100"), "0" -> "1" + "110010" + "0" = "11100100".
#   Components for "11011000" is ["11100100"].
#   Sort reverse: ["11100100"].
#   Join: "11100100".
#
# Time Complexity:
# The recursion depth can be up to N/2 (e.g., "101010...").
# At each level of recursion, we iterate through the string portion.
# The sorting of components also contributes. In the worst case, a string of length N might be broken into N/2 components of length 2. Sorting these N/2 strings of length 2 takes O((N/2) * 2 * log(N/2)) = O(N log N).
# However, the total length of strings processed across all recursive calls sums up.
# Each character is visited a constant number of times at each level of recursion. The number of levels can be O(N).
# The sorting step is crucial. If we have `k` components at a certain level, sorting them takes `O(k * L * log k)` where `L` is the max length of a component.
# In the worst case, the string can be decomposed into `O(N)` components.
# The overall time complexity is difficult to pin down precisely without a deeper analysis of the decomposition and sorting. However, it's often cited as roughly O(N^2 log N) or O(N^2) due to string concatenations and sorting.
# A more rigorous analysis might show it's closer to O(N^2) because the total length of all strings processed across all recursive calls is manageable, and the sorting of a fixed number of substrings within each level is dominant.
# Given N <= 50, an O(N^2) or O(N^2 log N) solution is acceptable.
#
# Space Complexity:
# The space complexity is dominated by the recursion depth and the storage of components.
# The maximum recursion depth can be O(N).
# At each recursive call, we create substrings and store them in a list. The total length of these substrings across all active recursive calls can be O(N^2) in a naive implementation, but the recursive structure ensures that the total size of strings being processed at any given time is manageable.
# The space for storing `components` at each level is proportional to the length of the current string segment.
# Thus, the space complexity is O(N) if we consider the recursion stack and O(N) for storing components at each level.
# String slicing and concatenation can also take space.
# Overall, it's often considered O(N) or O(N^2) depending on how string manipulations are accounted for. For N <= 50, this is fine.
class Solution:
    def makeLargestSpecial(self, s: str) -> str:
        # Base case: if the string is empty or too short to be special, return it.
        if not s or len(s) < 2:
            return s

        # List to store the special binary string components.
        components = []
        # Balance counter to identify complete special substrings.
        balance = 0
        # Start index for the current component.
        start_index = 0

        # Iterate through the string to find and process special substrings.
        for i, char in enumerate(s):
            # Increment balance for '1', decrement for '0'.
            if char == '1':
                balance += 1
            else:
                balance -= 1

            # When balance returns to 0, we've found a complete special substring.
            if balance == 0:
                # The current special substring is from start_index to i (inclusive).
                # The inner part of this special string (excluding the outer '1' and '0')
                # is s[start_index + 1 : i].
                # We recursively call makeLargestSpecial on this inner part to get
                # its lexicographically largest form.
                # Then we reconstruct the current special substring as '1' + largest_inner + '0'.
                inner_largest = self.makeLargestSpecial(s[start_index + 1 : i])
                components.append('1' + inner_largest + '0')
                # Update start_index for the next component.
                start_index = i + 1

        # Sort the found components in reverse lexicographical order.
        # This is the key step for making the overall string lexicographically largest.
        components.sort(reverse=True)

        # Join the sorted components to form the final largest special string.
        return "".join(components)

```